test/auto-languageclient.test.ts

```diff
@@ -1,25 +1,106 @@
+  }
+}
 
 describe("AutoLanguageClient", () => {
+  describe("ServerManager", () => {
+    describe("WorkspaceFolders", () => {
+      let client: FakeAutoLanguageClient
+      let serverManager: ServerManager
+      beforeEach(() => {
+        client = new FakeAutoLanguageClient()
+        client.activate()
+
+        /* eslint-disable-next-line dot-notation */
+        serverManager = client["_serverManager"]
+      })
+
+      describe("getWorkspaceFolders", () => {
+        it("returns null when no server is running", async () => {
+          const workspaceFolders = await serverManager.getWorkspaceFolders()
+          expect(workspaceFolders).to.be.null
+        })
+        it("returns null when a single file is open", async () => {
+          await atom.workspace.open(__filename)
+          const workspaceFolders = await serverManager.getWorkspaceFolders()
+          expect(workspaceFolders).to.be.null
+        })
+        it("gives the open workspace folders", async () => {
+          const projectPath = __dirname
+          const projectPath2 = dirname(__dirname)
+
+          const workspaceFolder = projectPathToWorkspaceFolder(normalizePath(projectPath))
+          const workspaceFolder2 = projectPathToWorkspaceFolder(normalizePath(projectPath2))
+
+          // gives the open workspace folder
+          atom.project.addPath(projectPath)
+          await serverManager.startServer(projectPath)
+          expect(await serverManager.getWorkspaceFolders()).to.deep.equals([workspaceFolder])
+
+          // doesn't give the workspace folder if it the server is not started for that project
+          atom.project.addPath(projectPath2)
+          expect(await serverManager.getWorkspaceFolders()).to.deep.equals([workspaceFolder])
+          await serverManager.startServer(projectPath)
+          expect(await serverManager.getWorkspaceFolders()).to.deep.equals([workspaceFolder, workspaceFolder2])
+        })
+      })
+      describe("didChangeWorkspaceFolders", () => {
+        it("gives a notification if the projects change", async () => {
+          const projectPath = __dirname
+          const projectPath2 = dirname(__dirname)
+
+          // const workspaceFolder = projectPathToWorkspaceFolder(normalizePath(projectPath))
+          const workspaceFolder2 = projectPathToWorkspaceFolder(normalizePath(projectPath2))
+
+          atom.project.addPath(projectPath)
+          await serverManager.startServer(projectPath)
+
+          // eslint-disable-next-line dot-notation
+          const server = serverManager["_activeServers"][0]
+
+          const spy = sinon.spy()
+          // eslint-disable-next-line dot-notation
+          server.connection.didChangeWorkspaceFolders = spy
+
+          atom.project.addPath(projectPath2)
+          console.log(spy.getCalls())
+
+          expect(spy.calledOnce).to.be.true
+          expect(spy.firstCall.args[0]).to.deep.equal({
+            added: [workspaceFolder2],
+            removed: []
+          })
```

Comment:
You forgot the `event` property.

```suggestion
          expect(spy.firstCall.args[0]).to.deep.equal({
            event: {
              added: [workspaceFolder2],
              removed: []
            }
          })
```

Reply:
also seems like maybe a race condition. If I run just this test it passes but running all tests makes this test fail.

Reply:
Thanks. I had to tear down the tests inside afterEach.